File: go-backend-interview-prep/10_design_patterns_in_go/08_singleton_and_why_avoid/main.go
package main

import (
	"fmt"
	"sync"
)

// ============================================================
// APPROACH 1: Singleton (the problem)
// ============================================================

type Config struct {
	SMTPHost string
	SMTPPort int
}

var (
	configInstance *Config
	configOnce     sync.Once
)

// GetConfig is a singleton -- returns the same instance every time.
func GetConfig() *Config {
	configOnce.Do(func() {
		fmt.Println("  [SINGLETON] Loading config (runs once)...")
		configInstance = &Config{SMTPHost: "smtp.example.com", SMTPPort: 587}
	})
	return configInstance
}

// singletonService uses the singleton directly -- hidden dependency!
type singletonService struct{}

func (s *singletonService) SendNotification(to, msg string) {
	cfg := GetConfig() // hidden: caller can't see this dependency
	fmt.Printf("  [SINGLETON] Send via %s:%d to=%s msg=%s\n",
		cfg.SMTPHost, cfg.SMTPPort, to, msg)
}

// ============================================================
// APPROACH 2: Dependency Injection (the solution)
// ============================================================

type Sender interface {
	Send(to, message string) error
}

type EmailSender struct {
	Host string
	Port int
}

func (e *EmailSender) Send(to, message string) error {
	fmt.Printf("  [DI]        Send via %s:%d to=%s msg=%s\n",
		e.Host, e.Port, to, message)
	return nil
}

type NotificationService struct {
	sender Sender // explicit dependency -- visible, testable
}

func NewNotificationService(s Sender) *NotificationService {
	return &NotificationService{sender: s}
}

func (n *NotificationService) Notify(to, msg string) {
	n.sender.Send(to, msg)
}

func main() {
	fmt.Println("=== Singleton and Why to Avoid It ===")
	fmt.Println()

	// --- Singleton approach ---
	fmt.Println("--- Singleton (hidden dependency) ---")
	svc1 := &singletonService{}
	svc1.SendNotification("alice@example.com", "Order shipped!")
	svc1.SendNotification("bob@example.com", "Payment received")
	fmt.Println()

	fmt.Println("  Problem: singletonService hides its Config dependency.")
	fmt.Println("  In tests, you can't swap Config without global mutation.")
	fmt.Println()

	// --- DI approach ---
	fmt.Println("--- DI alternative (explicit dependency) ---")
	sender := &EmailSender{Host: "smtp.example.com", Port: 587}
	svc2 := NewNotificationService(sender)
	svc2.Notify("alice@example.com", "Order shipped!")
	svc2.Notify("bob@example.com", "Payment received")
	fmt.Println()

	// In tests: swap with a fake
	fmt.Println("--- In tests: swap with fake ---")
	fakeSender := &FakeSender{}
	testSvc := NewNotificationService(fakeSender)
	testSvc.Notify("test@example.com", "test message")
	fmt.Printf("  Fake recorded: %v\n", fakeSender.calls)
	fmt.Println()

	fmt.Println("Key: DI makes dependencies visible and testable.")
	fmt.Println("     Singleton hides them. Prefer DI in almost all cases.")
}

type FakeSender struct {
	calls []string
}

func (f *FakeSender) Send(to, message string) error {
	f.calls = append(f.calls, to+":"+message)
	return nil
}

File: go-backend-interview-prep/10_design_patterns_in_go/08_singleton_and_why_avoid/main.py
"""Singleton and why to avoid it -- Python equivalent of the Go example.

Shows module-level singleton vs dependency injection alternative.
"""


# ============================================================
# APPROACH 1: Singleton (the problem)
# ============================================================

_config = None


def get_config():
    """Module-level singleton -- returns same dict every time."""
    global _config
    if _config is None:
        print("  [SINGLETON] Loading config (runs once)...")
        _config = {"smtp_host": "smtp.example.com", "smtp_port": 587}
    return _config


class SingletonService:
    """Uses the singleton directly -- hidden dependency!"""

    def send_notification(self, to: str, msg: str) -> None:
        cfg = get_config()  # hidden: caller can't see this dependency
        print(f"  [SINGLETON] Send via {cfg['smtp_host']}:{cfg['smtp_port']} to={to} msg={msg}")


# ============================================================
# APPROACH 2: Dependency Injection (the solution)
# ============================================================

class EmailSender:
    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port

    def send(self, to: str, message: str) -> None:
        print(f"  [DI]        Send via {self.host}:{self.port} to={to} msg={message}")


class NotificationService:
    def __init__(self, sender):
        self.sender = sender  # explicit dependency

    def notify(self, to: str, msg: str) -> None:
        self.sender.send(to, msg)


class FakeSender:
    def __init__(self):
        self.calls: list[str] = []

    def send(self, to: str, message: str) -> None:
        self.calls.append(f"{to}:{message}")


def main() -> None:
    print("=== Singleton and Why to Avoid It ===")
    print()

    # Singleton approach
    print("--- Singleton (hidden dependency) ---")
    svc1 = SingletonService()
    svc1.send_notification("alice@example.com", "Order shipped!")
    svc1.send_notification("bob@example.com", "Payment received")
    print()

    print("  Problem: SingletonService hides its config dependency.")
    print("  In tests, you can't swap config without global mutation.")
    print()

    # DI approach
    print("--- DI alternative (explicit dependency) ---")
    sender = EmailSender("smtp.example.com", 587)
    svc2 = NotificationService(sender)
    svc2.notify("alice@example.com", "Order shipped!")
    svc2.notify("bob@example.com", "Payment received")
    print()

    # In tests: swap with fake
    print("--- In tests: swap with fake ---")
    fake = FakeSender()
    test_svc = NotificationService(fake)
    test_svc.notify("test@example.com", "test message")
    print(f"  Fake recorded: {fake.calls}")
    print()

    print("Key: DI makes dependencies visible and testable.")
    print("     Singleton hides them. Prefer DI in almost all cases.")


if __name__ == "__main__":
    main()
